daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionConversionsSpec.scala

```diff
@@ -89,6 +94,268 @@ class TransactionConversionsSpec extends AnyWordSpec with Matchers {
       )
     }
   }
+
+  "decodeContractIdsAndKeys" should {
+    "return a single output for a create without a key" in {
+      val builder = TransactionBuilder()
+      val createNode = create(builder, "#1")
+      builder.add(createNode)
+      val result = TransactionConversions.decodeContractIdsAndKeys(
+        TransactionConversions
+          .encodeTransaction(builder.build())
+          .fold(error => fail(error.errorMessage), identity)
+      )
+      result shouldBe Right(Set(ContractIdOrKey.Id(createNode.coid)))
+    }
+
+    "return two outputs for a create with a key" in {
+      val builder = TransactionBuilder()
+      val createNode = create(builder, "#2", hasKey = true)
+      builder.add(createNode)
+      val result = TransactionConversions.decodeContractIdsAndKeys(
+        TransactionConversions
+          .encodeTransaction(builder.build())
+          .fold(error => fail(error.errorMessage), identity)
+      )
+      result shouldBe Right(
+        Set(
+          ContractIdOrKey.Id(createNode.coid),
+          ContractIdOrKey.Key(aGlobalKey),
+        )
+      )
+    }
+
+    "return a single output for a transient contract" in {
+      val builder = TransactionBuilder()
+      val createNode = create(builder, "#3", hasKey = true)
+      builder.add(createNode)
+      builder.add(exercise(builder, "#3"))
+      val result = TransactionConversions.decodeContractIdsAndKeys(
+        TransactionConversions
+          .encodeTransaction(builder.build())
+          .fold(error => fail(error.errorMessage), identity)
+      )
+      result shouldBe Right(
+        Set(
+          ContractIdOrKey.Id(createNode.coid),
+          ContractIdOrKey.Key(aGlobalKey),
+        )
+      )
+    }
+
+    "return a single output for an exercise without a key" in {
+      val builder = TransactionBuilder()
+      val exerciseNode = exercise(builder, "#4")
+      builder.add(exerciseNode)
+      val result = TransactionConversions.decodeContractIdsAndKeys(
+        TransactionConversions
+          .encodeTransaction(builder.build())
+          .fold(error => fail(error.errorMessage), identity)
+      )
+      result shouldBe Right(Set(ContractIdOrKey.Id(exerciseNode.targetCoid)))
+    }
+
+    "return two outputs for an exercise with a key" in {
```

Comment:
maybe worth adding a test for consuming vs non-consuming?

Reply:
Added, however not sure, as we don't treat them differently. Maybe will help someone in the future :)